File: services/communication-service/app/schemas/notification.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class NotificationRecipientWrite(BaseModel):
    recipientUserUuid: str
    recipientEmail: str = Field(..., min_length=1, max_length=255)
    recipientName: str = Field(..., min_length=1, max_length=255)


class NotificationCreateRequest(BaseModel):
    notificationType: str = Field(..., min_length=1, max_length=100)
    title: str = Field(..., min_length=1, max_length=200)
    body: str = Field(..., min_length=1)
    targetType: str | None = Field(default=None, max_length=100)
    targetId: str | None = Field(default=None, max_length=191)
    metadataJson: dict[str, Any] | None = None
    recipients: list[NotificationRecipientWrite] = Field(..., min_length=1)


class SystemNotificationRecipientWrite(BaseModel):
    recipientUserId: int = Field(..., ge=1)
    recipientEmail: str | None = Field(default=None, max_length=255)
    recipientName: str | None = Field(default=None, max_length=255)


class SystemNotificationCreateRequest(BaseModel):
    notificationType: str = Field(..., min_length=1, max_length=100)
    title: str = Field(..., min_length=1, max_length=200)
    body: str = Field(..., min_length=1)
    targetType: str | None = Field(default=None, max_length=100)
    targetId: str | None = Field(default=None, max_length=191)
    metadataJson: dict[str, Any] | None = None
    dedupeKey: str | None = Field(default=None, min_length=1, max_length=255)
    recipients: list[SystemNotificationRecipientWrite] = Field(..., min_length=1)


class NotificationUpdateRequest(BaseModel):
    notificationType: str | None = Field(default=None, min_length=1, max_length=100)
    title: str | None = Field(default=None, min_length=1, max_length=200)
    body: str | None = Field(default=None, min_length=1)
    targetType: str | None = Field(default=None, max_length=100)
    targetId: str | None = Field(default=None, max_length=191)
    metadataJson: dict[str, Any] | None = None


class NotificationRead(BaseModel):
    notificationId: int
    notificationUuid: str
    actorUserId: int | None
    actorUserUuid: str | None
    actorEmail: str | None
    actorName: str | None
    notificationType: str
    title: str
    body: str
    targetType: str | None
    targetId: str | None
    metadataJson: dict[str, Any] | None
    createdAt: datetime
    updatedAt: datetime


class PaginatedNotificationResponse(BaseModel):
    items: list[NotificationRead]
    page: int
    pageSize: int
    total: int
    totalPages: int


class NotificationRecipientRead(NotificationRead):
    recipientUserId: int
    recipientUserUuid: str
    recipientEmail: str
    recipientName: str
    isRead: bool
    readAt: datetime | None
    isHidden: bool
    hiddenAt: datetime | None
    receivedAt: datetime


class PaginatedNotificationRecipientResponse(BaseModel):
    items: list[NotificationRecipientRead]
    page: int
    pageSize: int
    total: int
    totalPages: int


class NotificationUnreadCountResponse(BaseModel):
    recipientUserId: int
    recipientUserUuid: str
    unreadCount: int


class NotificationRecipientStateResponse(BaseModel):
    notificationId: int
    notificationUuid: str
    recipientUserId: int
    recipientUserUuid: str
    isRead: bool
    readAt: datetime | None
    isHidden: bool
    hiddenAt: datetime | None


class MarkAllNotificationsReadResponse(BaseModel):
    recipientUserId: int
    recipientUserUuid: str
    updatedCount: int

File: frontend/src/types/notification.ts
export interface NotificationRecipientWrite {
  recipientUserUuid: string;
  recipientEmail: string;
  recipientName: string;
}

export interface NotificationCreateRequest {
  notificationType: string;
  title: string;
  body: string;
  targetType?: string | null;
  targetId?: string | null;
  metadataJson?: Record<string, unknown> | null;
  recipients: NotificationRecipientWrite[];
}

export interface NotificationUpdateRequest {
  notificationType?: string;
  title?: string;
  body?: string;
  targetType?: string | null;
  targetId?: string | null;
  metadataJson?: Record<string, unknown> | null;
}

export interface NotificationRead {
  notificationId: number;
  notificationUuid: string;
  actorUserId: number | null;
  actorUserUuid: string | null;
  actorEmail: string | null;
  actorName: string | null;
  notificationType: string;
  title: string;
  body: string;
  targetType: string | null;
  targetId: string | null;
  metadataJson: Record<string, unknown> | null;
  createdAt: string;
  updatedAt: string;
}

export interface PaginatedNotificationResponse {
  items: NotificationRead[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
}

export interface NotificationRecipientRead extends NotificationRead {
  recipientUserId: number;
  recipientUserUuid: string;
  recipientEmail: string;
  recipientName: string;
  isRead: boolean;
  readAt: string | null;
  isHidden: boolean;
  hiddenAt: string | null;
  receivedAt: string;
}

export interface PaginatedNotificationRecipientResponse {
  items: NotificationRecipientRead[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
}

export interface NotificationUnreadCountResponse {
  recipientUserId: number;
  recipientUserUuid: string;
  unreadCount: number;
}

export interface NotificationRecipientStateResponse {
  notificationId: number;
  notificationUuid: string;
  recipientUserId: number;
  recipientUserUuid: string;
  isRead: boolean;
  readAt: string | null;
  isHidden: boolean;
  hiddenAt: string | null;
}

export interface MarkAllNotificationsReadResponse {
  recipientUserId: number;
  recipientUserUuid: string;
  updatedCount: number;
}

export interface NotificationRecipientListParams {
  includeHidden?: boolean;
  unreadOnly?: boolean;
  notificationType?: string;
  page?: number;
  pageSize?: number;
}

export interface NotificationListParams {
  notificationType?: string;
  page?: number;
  pageSize?: number;
}
